### DISCUSSION TITLE
Full shader for new presets on Android

### DISCUSSION THREAD

**Discussion by OfficialIncubo** [projectM-visualizer/projectm#560] [Q&A] [answered]
Can you please implement the new preset shaders, such as "martin - cascade.milk, martin - colorwall.milk, Martin - Hades Disqoteque (Headache Edition).milk etc" for Android?

**Comment by kblaschke** [comment, marked as answer]
The shaders seem to be fine in these presets, but the presets all use currently unimplemented equation features like loops and megabuf, which will cause the preset to render nothing.

projectM will require a rewrite of the equation parser to be able to support these new functions, as the current parser can only work with `var  = eqn` assignments and does not support functions at the start of a code line.

**Reply by kblaschke** [reply]
I did lots of work on that recently, so you may try the new code in the currently open PR here:
https://github.com/projectM-visualizer/projectm/pull/716

The [equation parser was rewritten from scratch](https://github.com/projectM-visualizer/projectm-eval), now fully supporting all Milkdrop expressions including all the undocumented and weird stuff you can do in them. The renderer was also completely overhauled, further improving the visuals to look more like Milkdrop. Still some issues, some which we can't really fix due to HLSL/GLSL incompatibilities, but only very few presets should be affected (like "martin - stormy sea.milk"). These presets can mostly be fixed by slightly changing the warp/comp shader code in the milk file.

**Comment by kblaschke** [comment]
As libprojectM has been refactored in almost every aspect, preset should now only rarely look broken, no matter which platform they run on.

Regarding Android, there's still the projectM app in the Play Store: note this uses a rather old projectM library, e,g. 3.1, and the author has mostly stopped maintaining it. While there's a slight chance he'll update it to use the 4.1 release in the future, there's nothing we can do from our end. Mailing Peter Sperl and asking him to do so might help though.
